File: apps/api/app/schemas/crypto_order_previews.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, field_serializer

from app.schemas.exchange_connections import ExchangeEnvironment, ExchangeReadinessVerdict

CryptoOrderPreviewStatus = Literal[
    "DRAFT",
    "VALIDATING",
    "RISK_REJECTED",
    "CONNECTION_NOT_READY",
    "BALANCE_INSUFFICIENT",
    "PREVIEW_REQUESTED",
    "PREVIEW_READY",
    "PREVIEW_FAILED",
    "EXPIRED",
    "CANCELLED",
]
CryptoOrderPreviewSide = Literal["BUY", "SELL"]
CryptoOrderPreviewOrderType = Literal["MARKET"]
CryptoOrderPreviewAmountCurrency = Literal["USD", "BTC"]
CryptoOrderPreviewGeneratedBy = Literal["operator", "system_recommendation"]
CryptoOrderPreviewRiskVerdict = Literal["approved_for_preview", "rejected", "blocked", "needs_refresh"]


class CryptoOrderPreviewCreateRequest(BaseModel):
    exchange_connection_id: UUID
    environment: ExchangeEnvironment
    product_id: str = Field(default="BTC-USD", min_length=1)
    side: CryptoOrderPreviewSide
    order_type: CryptoOrderPreviewOrderType = "MARKET"
    quote_size: Decimal | None = None
    base_size: Decimal | None = None
    requested_amount_currency: CryptoOrderPreviewAmountCurrency = "USD"
    decision_record_id: UUID | None = None
    validation_run_id: UUID | None = None
    strategy_id: UUID | None = None
    strategy_name: str | None = None
    generated_by: CryptoOrderPreviewGeneratedBy = "operator"
    client_request_id: str | None = None


class CryptoOrderPreviewRefreshRequest(BaseModel):
    client_request_id: str | None = None


class CryptoOrderPreviewCancelRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=200)


class CryptoOrderPreviewReadinessResponse(BaseModel):
    ready: bool
    allowed_products: list[str]
    max_quote_size_usd: Decimal
    default_quote_size_usd: Decimal
    market_data_max_age_minutes: int
    expiration_minutes: int

    @field_serializer("max_quote_size_usd", "default_quote_size_usd", when_used="json")
    def serialize_decimal_fields(self, value: Decimal) -> str:
        return format(value, "f")


class CryptoOrderPreviewResponse(BaseModel):
    crypto_order_preview_id: UUID
    preview_version: int
    status: CryptoOrderPreviewStatus
    provider: str
    environment: ExchangeEnvironment
    product_id: str
    side: CryptoOrderPreviewSide
    order_type: CryptoOrderPreviewOrderType
    quote_size: Decimal | None
    base_size: Decimal | None
    requested_amount: Decimal
    requested_amount_currency: CryptoOrderPreviewAmountCurrency
    readiness_verdict: ExchangeReadinessVerdict | None
    risk_verdict: CryptoOrderPreviewRiskVerdict | None
    risk_explanation: str | None
    strategy_id: UUID | None
    strategy_name: str | None
    decision_record_id: UUID | None
    validation_run_id: UUID | None
    preview_id: str | None
    estimated_average_price: Decimal | None
    estimated_total_value: Decimal | None
    estimated_base_size: Decimal | None
    estimated_quote_size: Decimal | None
    estimated_fee: Decimal | None
    estimated_fee_currency: str | None
    estimated_slippage: Decimal | None
    estimated_commission_total: Decimal | None
    best_bid: Decimal | None
    best_ask: Decimal | None
    available_balance_before: Decimal | None
    estimated_balance_after: Decimal | None
    failure_reason: str | None
    warning_messages: list[str]
    exchange_response_summary: dict[str, object]
    expires_at: datetime
    generated_by: CryptoOrderPreviewGeneratedBy
    audit_correlation_id: UUID | None
    order_submitted: bool
    execution_available: bool
    created_at: datetime
    updated_at: datetime
    refreshed_from_preview_id: UUID | None = None

    @field_serializer(
        "quote_size",
        "base_size",
        "requested_amount",
        "estimated_average_price",
        "estimated_total_value",
        "estimated_base_size",
        "estimated_quote_size",
        "estimated_fee",
        "estimated_slippage",
        "estimated_commission_total",
        "best_bid",
        "best_ask",
        "available_balance_before",
        "estimated_balance_after",
        when_used="json",
    )
    def serialize_decimals(self, value: Decimal | None) -> str | None:
        if value is None:
            return None
        return format(value, "f")


class CryptoOrderPreviewListResponse(BaseModel):
    items: list[CryptoOrderPreviewResponse]


class CryptoOrderPreviewDetailResponse(CryptoOrderPreviewResponse):
    pass

File: apps/web/lib/api/crypto-order-previews.ts
import { ApiRequestError } from "@/lib/api/arena";
import { getOperatorAuthHeaders } from "@/lib/api/operator-auth";

export type CryptoOrderPreviewStatus =
  | "DRAFT"
  | "VALIDATING"
  | "RISK_REJECTED"
  | "CONNECTION_NOT_READY"
  | "BALANCE_INSUFFICIENT"
  | "PREVIEW_REQUESTED"
  | "PREVIEW_READY"
  | "PREVIEW_FAILED"
  | "EXPIRED"
  | "CANCELLED";

export type CryptoOrderPreviewSide = "BUY" | "SELL";
export type CryptoOrderPreviewOrderType = "MARKET";
export type CryptoOrderPreviewGeneratedBy = "operator" | "system_recommendation";
export type CryptoOrderPreviewReadinessVerdict =
  | "READY_FOR_PREVIEW"
  | "READ_ONLY_READY"
  | "MISCONFIGURED"
  | "UNREACHABLE"
  | "PERMISSION_INSUFFICIENT"
  | "CLOCK_SKEW"
  | "AUTHENTICATION_FAILED"
  | "UNKNOWN";
export type CryptoOrderPreviewRiskVerdict = "approved_for_preview" | "rejected" | "blocked" | "needs_refresh";

export type CryptoOrderPreviewReadiness = {
  ready: boolean;
  allowed_products: string[];
  max_quote_size_usd: string;
  default_quote_size_usd: string;
  market_data_max_age_minutes: number;
  expiration_minutes: number;
};

export type CryptoOrderPreview = {
  crypto_order_preview_id: string;
  preview_version: number;
  status: CryptoOrderPreviewStatus;
  provider: string;
  environment: "sandbox" | "production";
  product_id: string;
  side: CryptoOrderPreviewSide;
  order_type: CryptoOrderPreviewOrderType;
  quote_size: string | null;
  base_size: string | null;
  requested_amount: string;
  requested_amount_currency: "USD" | "BTC";
  readiness_verdict: CryptoOrderPreviewReadinessVerdict | null;
  risk_verdict: CryptoOrderPreviewRiskVerdict | null;
  risk_explanation: string | null;
  strategy_id: string | null;
  strategy_name: string | null;
  decision_record_id: string | null;
  validation_run_id: string | null;
  preview_id: string | null;
  estimated_average_price: string | null;
  estimated_total_value: string | null;
  estimated_base_size: string | null;
  estimated_quote_size: string | null;
  estimated_fee: string | null;
  estimated_fee_currency: string | null;
  estimated_slippage: string | null;
  estimated_commission_total: string | null;
  best_bid: string | null;
  best_ask: string | null;
  available_balance_before: string | null;
  estimated_balance_after: string | null;
  failure_reason: string | null;
  warning_messages: string[];
  exchange_response_summary: Record<string, unknown>;
  expires_at: string;
  generated_by: CryptoOrderPreviewGeneratedBy;
  audit_correlation_id: string | null;
  order_submitted: boolean;
  execution_available: boolean;
  created_at: string;
  updated_at: string;
  refreshed_from_preview_id: string | null;
};

export type CryptoOrderPreviewCreateRequest = {
  exchange_connection_id: string;
  environment: "sandbox" | "production";
  product_id: string;
  side: CryptoOrderPreviewSide;
  order_type: CryptoOrderPreviewOrderType;
  quote_size?: string | null;
  base_size?: string | null;
  requested_amount_currency: "USD" | "BTC";
  decision_record_id?: string | null;
  validation_run_id?: string | null;
  strategy_id?: string | null;
  strategy_name?: string | null;
  generated_by?: CryptoOrderPreviewGeneratedBy;
  client_request_id?: string | null;
};

export type CryptoOrderPreviewRefreshRequest = {
  client_request_id?: string | null;
};

export type CryptoOrderPreviewCancelRequest = {
  reason: string;
};

type ErrorEnvelope = {
  error?: {
    code?: string;
    message?: string;
    details?: Record<string, unknown>;
  };
};

const API_BASE_URL = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

async function requestJson<T>(path: string, init?: RequestInit): Promise<T> {
  const response = await fetch(`${API_BASE_URL}${path}`, {
    cache: "no-store",
    headers: {
      "Content-Type": "application/json",
      ...getOperatorAuthHeaders(),
      ...(init?.headers ?? {}),
    },
    ...init,
  });

  if (!response.ok) {
    let message = `Request failed with status ${response.status}`;
    let code: string | undefined;
    try {
      const payload = (await response.json()) as ErrorEnvelope;
      code = payload.error?.code;
      if (payload.error?.message) {
        message = payload.error.message;
      }
    } catch {
      // Keep generic message.
    }
    throw new ApiRequestError(message, response.status);
  }

  return (await response.json()) as T;
}

export async function getCryptoOrderPreviewReadiness(): Promise<CryptoOrderPreviewReadiness> {
  return requestJson<CryptoOrderPreviewReadiness>("/crypto-order-previews/readiness");
}

export async function listCryptoOrderPreviews(limit = 25): Promise<CryptoOrderPreview[]> {
  return requestJson<{ items: CryptoOrderPreview[] }>(`/crypto-order-previews?limit=${encodeURIComponent(String(limit))}`)
    .then((payload) => payload.items);
}

export async function getCryptoOrderPreview(previewId: string): Promise<CryptoOrderPreview> {
  return requestJson<CryptoOrderPreview>(`/crypto-order-previews/${encodeURIComponent(previewId)}`);
}

export async function createCryptoOrderPreview(payload: CryptoOrderPreviewCreateRequest): Promise<CryptoOrderPreview> {
  return requestJson<CryptoOrderPreview>("/crypto-order-previews", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

export async function refreshCryptoOrderPreview(previewId: string, payload?: CryptoOrderPreviewRefreshRequest): Promise<CryptoOrderPreview> {
  return requestJson<CryptoOrderPreview>(`/crypto-order-previews/${encodeURIComponent(previewId)}/refresh`, {
    method: "POST",
    body: JSON.stringify(payload ?? {}),
  });
}

export async function cancelCryptoOrderPreview(previewId: string, payload: CryptoOrderPreviewCancelRequest): Promise<CryptoOrderPreview> {
  return requestJson<CryptoOrderPreview>(`/crypto-order-previews/${encodeURIComponent(previewId)}/cancel`, {
    method: "POST",
    body: JSON.stringify(payload),
  });
}
